### PR TITLE
Add a docker file

### PR DESCRIPTION
**Pull request by halkeye**
Just a quick docker file so an official image can be setup (its pretty easy to do on hub.docker, but I assume as a go dev you've played with it before).

I like to run everything out of docker/kubernetes, so figured I'd help you make an "official" image.

I'll be making a helm chart down the road, and would love to see actual tagged releases, but /shrug

### CODE REVIEW

**Comment by halkeye** [comment]
```  master ?  ~/git/irc-slack  docker build -t gavin .
Sending build context to Docker daemon    384kB
Step 1/9 : FROM golang:alpine AS builder
 ---> f56365ec0638
Step 2/9 : RUN apk update && apk add --no-cache git
 ---> Using cache
 ---> 2586c782c509
Step 3/9 : COPY . $GOPATH/src/insomniacslk/irc-slack
 ---> Using cache
 ---> eda2de686501
Step 4/9 : WORKDIR $GOPATH/src/insomniacslk/irc-slack
 ---> Using cache
 ---> 3160fc42e69e
Step 5/9 : RUN go get -d -v
 ---> Using cache
 ---> 84a15ce26dd8
Step 6/9 : RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -installsuffix cgo -ldflags="-w -s" -o /go/bin/irc-slack
 ---> Running in 5936a99a8e61
Removing intermediate container 5936a99a8e61
 ---> 78a775f9316f
Step 7/9 : FROM scratch
 ---> 
Step 8/9 : COPY --from=builder /go/bin/irc-slack /go/bin/irc-slack
 ---> 58536a5b8e7f
Step 9/9 : ENTRYPOINT ["/go/bin/irc-slack"]
 ---> Running in 6de8c5987161
Removing intermediate container 6de8c5987161
 ---> d223885742d4
Successfully built d223885742d4
Successfully tagged gavin:latest
Time: 0h:00m:14s               ```


```
  add-docker-file ?  ~/git/irc-slack  docker run --rm -it gavin --help
Usage of /go/bin/irc-slack:
  -H string
        IP address to listen on (default "127.0.0.1")
  -chunk int
        Maximum size of a line to send to the client. Only works for certain reply types (default 512)
  -p int
        Local port to listen on (default 6666)
  -s string
        IRC server name (i.e. the host name to send to clients)

```

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# Install git.
+# Git is required for fetching the dependencies.
+RUN apk update && apk add --no-cache git
+COPY . $GOPATH/src/insomniacslk/irc-slack
+WORKDIR $GOPATH/src/insomniacslk/irc-slack
+# Fetch dependencies.
+# Using go get.
+RUN go get -d -v
+# Build the binary.
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -installsuffix cgo -ldflags="-w -s" -o /go/bin/irc-slack
```
Comment (insomniacslk) [inline Dockerfile]: Why forcing it to `amd64`? I am not familiar with Docker but I assume that people may want to build it for other architectures as well (e.g. ARM, to run on a RPi).

Also, `installsuffix` is not required, see https://github.com/golang/go/issues/9344#issuecomment-69944514

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+############################
+# STEP 2 build a small image
+############################
+FROM scratch
+# Copy our static executable.
+COPY --from=builder /go/bin/irc-slack /go/bin/irc-slack
```
Comment (insomniacslk) [inline Dockerfile]: Same source and destination, is this really required? Unless this is something that in Docker's language means something else?

Comment (halkeye) [reply Dockerfile]: its copying from the first container (build container)'s /go/bin/irc-slack to the empty container /go/bin/irc-slack

So yea I think its necessary, i don't end up using the multi container stuff very often myself so not 100%

File: Dockerfile (hunk 1 above)
Comment (halkeye) [reply Dockerfile]: ah right, i was just copy and pasting from a tutorial, was lazy to type it out all out.

I'll remove it

Comment (halkeye) [reply Dockerfile]: I never remember the CGO stuff offhand, so always copy it from online, i should just start typing it out and remembering it

**Comment by halkeye** [comment]
> I guess this requires uploading the dockerfile to the docker hub and pulling/building it via Travis-CI?

good question. I've never used dockerhub build infra when using regular CI, you can just push to docker hub.

https://docs.travis-ci.com/user/docker/#pushing-a-docker-image-to-a-registry

**Comment by halkeye** [comment]
Added an attempt to travis.yml

**Comment by halkeye** [comment]
All tested now! had so many minor issues, its been a while since I setup travis

https://travis-ci.org/halkeye/irc-slack/branches

**Comment by halkeye** [comment]
sweet! and it'll do tagged releases if/when you do them!

https://hub.docker.com/r/halkeye/irc-slack/tags

**Comment by insomniacslk** [comment]
Thanks for all this work! The PR looks good to me. I guess that before merging it I should follow the steps at https://docs.travis-ci.com/user/docker/#pushing-a-docker-image-to-a-registry to use the credentials I created at hub.docker.com, is that correct?

**Comment by halkeye** [comment]
yep, mostly just:

travis env set DOCKER_USERNAME myusername
travis env set DOCKER_PASSWORD secretsecret

Then create the repo (which it looks like you did) for https://hub.docker.com/r/insomniacslk/irc-slack

I thought there was a way to generate a token for docker, but I can't find it now, so you have to use your dockerhub password

**Comment by insomniacslk** [comment]
it seems like I could set those variables via Travis's UI rather than in .travis.yml, I'll merge this now

**Comment by insomniacslk** [comment]
Thanks for your contribution @halkeye !

**Pull request merged**